src/chainstate/stacks/db/blocks.rs

```diff
@@ -1130,169 +1056,337 @@ impl StacksChainState {
+    /// If processed_only is true, then only processed microblocks are loaded
+    fn inner_load_microblock_stream_fork(
+        blocks_conn: &DBConn,
+        parent_consensus_hash: &ConsensusHash,
+        parent_anchored_block_hash: &BlockHeaderHash,
+        tip_microblock_hash: &BlockHeaderHash,
+        processed_only: bool,
+    ) -> Result<Option<Vec<StacksMicroblock>>, Error> {
+        let mut ret = vec![];
+        let mut mblock_hash = tip_microblock_hash.clone();
+        let mut last_seq = u16::MAX;
 
-        // merge staging and on-disk streams
-        let mut microblocks = vec![];
-        for i in 0..cnt {
-            // favor DB-stored microblock over disk-stored, since the DB is less likely to be
-            // corrupt!
-            if !staging_microblocks[i].processed {
-                if staging_microblocks[i].block_data.len() == 0 {
-                    return Err(Error::NetError(net_error::DeserializeError(format!(
-                        "Microblock {} does not have block data",
-                        staging_microblocks[i].microblock_hash
-                    ))));
-                }
+        loop {
+            let microblock =
+                match StacksChainState::load_staging_microblock_bytes(blocks_conn, &mblock_hash)? {
+                    Some(mblock_data) => StacksMicroblock::consensus_deserialize(
+                        &mut &mblock_data[..],
+                    )
+                    .expect(&format!(
+                        "CORRUPTION: failed to parse microblock data for {}/{}-{}",
+                        parent_consensus_hash, parent_anchored_block_hash, &mblock_hash,
+                    )),
+                    None => {
+                        debug!(
+                            "No such microblock (processed={}): {}/{}-{} ({})",
+                            processed_only,
+                            parent_consensus_hash,
+                            parent_anchored_block_hash,
+                            &mblock_hash,
+                            last_seq
+                        );
+                        return Ok(None);
+                    }
+                };
 
-                let microblock = StacksMicroblock::consensus_deserialize(
-                    &mut &staging_microblocks[i].block_data[..],
-                )
-                .map_err(Error::NetError)?;
-                microblocks.push(microblock);
-            } else {
-                microblocks.push(disk_microblocks[i].clone());
+            test_debug!(
+                "Loaded microblock {}/{}-{} (parent={}, expect_seq={})",
+                &parent_consensus_hash,
+                &parent_anchored_block_hash,
+                &microblock.block_hash(),
+                &microblock.header.prev_block,
+                last_seq.saturating_sub(1)
+            );
+
+            if last_seq < u16::MAX
+                && microblock.header.sequence < u16::MAX
+                && microblock.header.sequence > 0
```

Comment:
Shouldn't this be true when `microblock.header.sequence == 0` as well?

Reply:
That line could be removed, since `sequence` is a `u16`

Reply:
Addressed in ee12847